src/statement.d

```diff
@@ -3839,48 +3839,64 @@ public:
         if (condition.op == TOKerror)
             conditionError = true;
 
-        bool needswitcherror = false;
+        cases = new CaseStatements();
 
         sc = sc.push();
         sc.sbreak = this;
         sc.sw = this;
-
-        cases = new CaseStatements();
         sc.noctor++; // BUG: should use Scope::mergeCallSuper() for each case instead
         _body = _body.semantic(sc);
         sc.noctor--;
+        sc = sc.pop();
 
         if (conditionError || _body.isErrorStatement())
-            goto Lerror;
+            return new ErrorStatement();
```

Comment:
The reason for the goto's are:
1. handy place to put a printf on function return. Having lots of return statements makes this a pain.
2. cuts the size of the code generated.